### PR TITLE
bpf: nodeport: integrate Ingress RevSNAT and RevDNAT paths

### PR DESCRIPTION
**Pull request by julianwiedmann**
The Ingress RevSNAT and RevDNAT handling is currently split into two separate tail-calls. Which means that if the RevSNAT code performs a CT lookup and receives trace information, then we can't easily pass the corresponding `trace_ctx` to the `TO_OVERLAY` trace event in the RevDNAT code.

In practice this means that when in direct-routing mode, the trace events for EgressGW reply traffic don't benefit from CT-driven aggregation.

This PR refactors the relevant code paths, and turns the RevDNAT path into a conditional tail-call. On newer kernels it gets inlined into `tail_nodeport_nat_ingress_ipv4()`, and we can pass through the `trace_ctx` struct all the way to the `send_trace_notify()`.

### CODE REVIEW

**Comment by julianwiedmann** [comment]
Sorry for the massive churn. But this needed a fair bit of refactoring to slash through all the tech debt :grimacing: .

File: bpf/lib/nodeport.h [hunk 1]
```diff
@@ -2452,16 +2451,17 @@ static __always_inline int rev_nodeport_lb4(struct __ctx_buff *ctx, __s8 *ext_er
...
 	int ret = 0;
 #if defined(ENABLE_HOST_FIREWALL) && defined(IS_BPF_HOST)
 	/* We only enforce the host policies if nodeport.h is included from
 	 * bpf_host.
 	 */
-	struct trace_ctx __maybe_unused trace = {
-		.reason = TRACE_REASON_UNKNOWN,
-		.monitor = 0,
-	};
+
```
Comment (aspsk) [inline bpf/lib/nodeport.h]: Extra line

File: bpf/lib/nodeport.h [hunk 2]
```diff
@@ -2544,15 +2567,29 @@ int tail_nodeport_nat_ingress_ipv4(struct __ctx_buff *ctx)
...
+					&trace, &ext_err);
+		if (IS_ERR(ret))
+			goto drop_err;
+
+		/* No redirect needed: */
+		if (ret == CTX_ACT_OK)
+			goto recircle;
+
+		/* Redirected to egress interface: */
+		edt_set_aggregate(ctx, 0);
```
Comment (aspsk) [inline bpf/lib/nodeport.h:2584]: previously we set it only when !LXC. Can this be called from LXC?

**Review by aspsk — COMMENTED** [review]
LGTM, left one question and one cosmetic comment

File: bpf/lib/nodeport.h (hunk 1 above)
Comment (julianwiedmann) [reply bpf/lib/nodeport.h]: nice catch, thanks! Fixed.

File: bpf/lib/nodeport.h (hunk 2 above)
Comment (julianwiedmann) [reply bpf/lib/nodeport.h:2584]: I added the following to the patch description, thank you!

>    Note that CILIUM_CALL_IPV*_NODEPORT_NAT_INGRESS is never invoked from  bpf_lxc, so we can remove the #ifdef's around edt_set_aggregate().

**Review by aspsk — APPROVED** [review]
LGTM, thanks!

**Comment by julianwiedmann** [comment]
> (one last rebase to pull #27509 in)

this now required a small change for this PR: https://github.com/cilium/cilium/pull/27488/commits/944acfd51aa4280bc21eac5221d69c81df6794c5.

**Pull request merged**